Add bottom dialog pointer to color picker and fix positioning

volfied opened the pull request:
Using [CSS Mixin Guards](http://lesscss.org/features/#mixin-guards-feature), we now position the dialog pointer either at the top or the bottom of color picker dialog.

Also, addressing #3550, we now position the color picker correctly when there is no room below the clicked color input and color picker appears above it.

volfied commented on src/style/panel.less:
```diff
@@ -383,7 +383,7 @@
     
     &:before {
-        .dialog-pointer(5.4rem);
+        .dialog-pointer(5.4rem, true);
```
I have to add this because there are no default parameters with less mixins.

volfied commented on src/style/shared/dialog.less:
```diff
@@ -190,10 +190,20 @@ dialog::backdrop {
 }
 
-.dialog-pointer(@left) {
+.pointerMixin (@atTop) when (@atTop) {
```
This is equivalent to `@atTop === true` in JS

volfied commented on src/style/shared/dialog.less:
```diff
@@ -190,10 +190,20 @@ dialog::backdrop {
+    transform: rotate(45deg);
+}
+
+.pointerMixin (@atTop) when (default()) {
```
This is all the cases not accounted for `.pointerMixin`

volfied commented on src/style/shared/dialog.less:
```diff
@@ -190,10 +190,20 @@ dialog::backdrop {
+}
+
+.dialog-pointer(@left, @atTop) {
+    .pointerMixin(@atTop);
```
And this is how you call a prepared mixin, adding the properties it calculates into the class.

iwehrman commented:
Here's a failure case involving a short window:
![image](https://cloud.githubusercontent.com/assets/1075154/12436886/ded3c334-becd-11e5-9623-542956778980.png)

iwehrman commented:
Note that in the case above I think it should just show neither pointer.

volfied commented:
@iwehrman thanks for the catch, I changed the class names to be more explicit and fixed the issue so it doesn't draw any pointers if it's going to float.

iwehrman commented on src/js/jsx/shared/Dialog.jsx:
```diff
@@ -358,8 +360,10 @@ define(function (require, exports, module) {
                 classes = classnames(globalClass, this.props.className || "",
-                    { "placed-above": this.state.placedAbove }),
+                    { "pointer-above": !this.state.pointerBelow },
+                    { "pointer-below": !this.state.pointerAbove }),
```
Hmm, this is kind of confusing. Is it possible for these two sets of names to be more in sync?

volfied replied:
Fixed it, thanks.

volfied commented:
Back to you, @iwehrman

iwehrman commented:
![image](https://cloud.githubusercontent.com/assets/1075154/12466125/fb89eaf8-bf86-11e5-9b97-e224cc8e5808.png)

volfied commented:
After a remote debugging session, we realized this issue happens when the swatch is half rendered outside the panel bounds (bottom one in the screenshot). Making a note here for myself.

iwehrman commented:
A fix for this may also address #3440.

volfied commented:
@iwehrman I can't reproduce this again, I've tried at all window sizes / color picker locations.

iwehrman commented:
Hmm, I also can't repro at the moment. I guess that means it's fixed!
